rcl/src/rcl/wait.c

```diff
@@ -0,0 +1,448 @@
+
+#define SET_RESIZE_RMW_DEALLOC(RMWStorage, RMWCount) \
+  /* Also deallocate the rmw storage. */ \
+  if (wait_set->impl->RMWStorage) { \
+    allocator.deallocate((void *)wait_set->impl->RMWStorage, allocator.state); \
+    wait_set->impl->RMWStorage = NULL; \
+  }
+
+#define SET_RESIZE_RMW_REALLOC(Type, RMWStorage, RMWCount) \
+  /* Also resize the rmw storage. */ \
+  wait_set->impl->RMWCount = 0; \
+  wait_set->impl->RMWStorage = (void **)allocator.reallocate( \
+    wait_set->impl->RMWStorage, sizeof(rcl_ ## Type ## _t *) * size, allocator.state); \
+  if (!wait_set->impl->RMWStorage) { \
+    allocator.deallocate((void *)wait_set->Type ## s, allocator.state); \
+    wait_set->size_of_ ## Type ## s = 0; \
+    RCL_SET_ERROR_MSG("allocating memory failed"); \
+    return RCL_RET_BAD_ALLOC; \
+  } \
+  wait_set->impl->RMWCount = size;
+
+rcl_ret_t
+rcl_wait_set_add_subscription(
+  rcl_wait_set_t * wait_set,
+  const rcl_subscription_t * subscription)
+{
+  SET_ADD(subscription)
+  SET_ADD_RMW(subscription, rmw_subscriptions.subscribers)
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_wait_set_clear_subscriptions(rcl_wait_set_t * wait_set)
+{
+  SET_CLEAR(subscription)
+  SET_CLEAR_RMW(
+    subscription,
+    rmw_subscriptions.subscribers,
+    rmw_subscriptions.subscriber_count)
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_wait_set_resize_subscriptions(rcl_wait_set_t * wait_set, size_t size)
+{
+  SET_RESIZE(
+    subscription,
+    SET_RESIZE_RMW_DEALLOC(
+      rmw_subscriptions.subscribers, rmw_subscriptions.subscriber_count),
+    SET_RESIZE_RMW_REALLOC(
+      subscription, rmw_subscriptions.subscribers, rmw_subscriptions.subscriber_count)
+  )
+}
+
+rcl_ret_t
+rcl_wait_set_add_guard_condition(
+  rcl_wait_set_t * wait_set,
+  const rcl_guard_condition_t * guard_condition)
+{
+  SET_ADD(guard_condition)
+  SET_ADD_RMW(guard_condition, rmw_guard_conditions.guard_conditions)
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_wait_set_clear_guard_conditions(rcl_wait_set_t * wait_set)
+{
+  SET_CLEAR(guard_condition)
+  SET_CLEAR_RMW(
+    guard_condition,
+    rmw_guard_conditions.guard_conditions,
+    rmw_guard_conditions.guard_condition_count)
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_wait_set_resize_guard_conditions(rcl_wait_set_t * wait_set, size_t size)
+{
+  SET_RESIZE(
+    guard_condition,
+    SET_RESIZE_RMW_DEALLOC(
+      rmw_guard_conditions.guard_conditions, rmw_guard_conditions.guard_condition_count),
+    SET_RESIZE_RMW_REALLOC(
+      guard_condition,
+      rmw_guard_conditions.guard_conditions,
+      rmw_guard_conditions.guard_condition_count)
+  )
+}
+
+rcl_ret_t
+rcl_wait_set_add_timer(
+  rcl_wait_set_t * wait_set,
+  const rcl_timer_t * timer)
+{
+  SET_ADD(timer)
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_wait_set_clear_timers(rcl_wait_set_t * wait_set)
+{
+  SET_CLEAR(timer)
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_wait_set_resize_timers(rcl_wait_set_t * wait_set, size_t size)
+{
+  SET_RESIZE(timer,;,;)  // NOLINT
```

Comment:
Yeah it's kind of nasty, basically it assumes size is defined in the local scope out side of the macro call.